### PR TITLE
Remove the flannel experimental overlay

### LINKED ISSUE #17795: Cleanup kubelet flannel hacks
Introduced in https://github.com/kubernetes/kubernetes/pull/17668.

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

It removes the deprecated flannel overlay integration in kubelet.
It's completely unnecessary now with CNI which can handle everything smoothly when flannel is running in a daemonset.

**Which issue this PR fixes** _(optional, in `fixes #<issue number>(, #<issue_number>, ...)` format, will close that issue when PR gets merged)_: fixes #17795, #30589

cc @kubernetes/sig-network @thockin @freehan @bprashanth @yujuhong @dchen1107

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33862)

### CODE REVIEW

**Comment by luxas** [comment]
I can also remove the proposal if you like

**Comment by yujuhong** [comment]
@luxas you need to run `hack/update-all.sh` to update the generated files.

**Comment by luxas** [comment]
Yes, I know :)
I just uploaded the first commit while hack/update-all.sh was running.

@yujuhong PTAL at the first commit

**Comment by yujuhong** [comment]
> I just uploaded the first commit while hack/update-all.sh was running.

Fair enough. It does take quite a while to run.

**Comment by luxas** [comment]
@yujuhong What do you think about the PR itself?

**Comment by yujuhong** [comment]
Salt needs some cleanup too: https://github.com/kubernetes/kubernetes/blob/51198f59da4cab0c2a17bafd09667e2ef7690fbd/cluster/saltbase/salt/kubelet/default#L107

**Comment by freehan** [comment]
@luxas Thanks for the PR! 

Just the quota exceeded. 
```
ERROR: (gcloud.compute.addresses.create) Some requests did not succeed:
 - Quota 'STATIC_ADDRESSES' exceeded. Limit: 20.0
```

**Comment by yujuhong** [comment]
> @yujuhong What do you think about the PR itself?

I'm fine with removing this flag, and in general happy to see dead code go. Someone from the cluster team may want to chime in.

**Comment by luxas** [comment]
Should I remove the old proposal as well: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/flannel-integration.md?

**Comment by luxas** [comment]
I'm testing now to remove the old flannel code in kube-up as well.

cc @roberthbailey @mikedanese @zmerlynn from the cluster team, PTAL

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -424,25 +424,25 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *Kub
...
-		podsPerCore:                int(kubeCfg.PodsPerCore),
-		nvidiaGPUs:                 int(kubeCfg.NvidiaGPUs),
-		syncLoopMonitor:            atomic.Value{},
-		resolverConfig:             kubeCfg.ResolverConfig,
-		cpuCFSQuota:                kubeCfg.CPUCFSQuota,
-		daemonEndpoints:            daemonEndpoints,
-		containerManager:           kubeDeps.ContainerManager,
-		nodeIP:                     net.ParseIP(kubeCfg.NodeIP),
-		clock:                      clock.RealClock{},
+		os:                kubeDeps.OSInterface,
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: why changing this?

**Comment by yujuhong** [comment]
> Should I remove the old proposal as well: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/flannel-integration.md?

I don't know the policy on proposals, but I think removing a no-longer-relevant proposal is good.

File: pkg/kubelet/kubelet_network.go [hunk 2]
```diff
@@ -121,7 +121,7 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
 	// is used or whether we are using overlay networking. We should return
 	// true for cloud providers if they implement Routes() interface and
 	// we are not using overlay networking.
-	if kl.cloud == nil || kl.cloud.ProviderName() != "gce" || kl.flannelExperimentalOverlay {
+	if kl.cloud == nil || kl.cloud.ProviderName() != "gce" {
```
Comment (mikedanese) [inline pkg/kubelet/kubelet_network.go]: we still want a flag that does this probably. flannelExperimentalOverlay is the wrong name.

**Comment by mikedanese** [comment]
> Should I remove the old proposal as well:

No. You can just leave it as is or make a not that it's deprecated.

File: pkg/kubelet/kubelet_network.go (hunk 2 above)
Comment (freehan) [inline pkg/kubelet/kubelet_network.go]: One can use the CNI flannel plugin. But that needs extra setup for flannel daemon and etc. 

I think that much cleaner than having a specific flag for flannel.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (luxas) [inline pkg/kubelet/kubelet.go]: It's just gofmt. I removed the two flannel overlay-specific fields

File: pkg/kubelet/kubelet_network.go (hunk 2 above)
Comment (luxas) [inline pkg/kubelet/kubelet_network.go]: At least, this isn't a regression. 
Nobody (at least not many) has used the experimental flannel overlay, and if we want to add a new flag for this specific case, we can do it later

Comment (mikedanese) [inline pkg/kubelet/kubelet_network.go]: @freehan I think we have the same problem for not just flannel. If I'm using the gce cloud provider with weave and I don't want to configure routes, kubelet will report not ready. There is no way of opting out of gce routes with the gce cloud provider. Is that an issue?

Comment (freehan) [inline pkg/kubelet/kubelet_network.go]: @mikedanese what does kubelet complain in the weave case? I do not recall kubelet checks podCIDR unless it comes from network plugin. I think the assigned route can be ignored by configuration.

Comment (luxas) [inline pkg/kubelet/kubelet_network.go]: I think you should open a separate issue for that discussion.
Please do that and keep focusing on the PR itself

**Comment by luxas** [comment]
@freehan Any concerns/comments?
If this looks good to you and @yujuhong I'll just rebase and make it ready to merge

**Comment by freehan** [comment]
LGTM 
Thanks!

**Comment by yujuhong** [comment]
LGTM. Please rebase.

**Comment by freehan** [comment]
Feel free to rebase and self apply

**Comment by luxas** [comment]
@k8s-bot unit test this

File: pkg/kubelet/kubelet_network.go (hunk 2 above)
Comment (dcbw) [inline pkg/kubelet/kubelet_network.go]: I brought this up in https://github.com/kubernetes/kubernetes/pull/33573#issuecomment-251719452.  It's a problem for OpenShift when run on GCE too, as we use an overlay and don't need Routes created for pod<->pod communication, but the NetworkUnavailable condition is set anyway.

But I think this should be solved differently and more cleanly than repurposing the flannelExperimentalOverlay flag.

**Pull request merged**